File: python/openvino-perf.py
import numpy as np
import torch
import argparse
import time
from main import load_image, build_dataset, MetricLogger, WARMUP_SEC, TEST_SEC
from timm.utils import accuracy
import openvino as ov
from openvino.runtime import get_version
from openvino.runtime.utils.types import get_dtype

def get_args_parser():
    parser = argparse.ArgumentParser(
        'EdgeTransformerPerf tflite_runtime evaluation and benchmark script', add_help=False)
    parser.add_argument('--extern-model', default=None, type=str, help='extern model name;resolution')
    parser.add_argument('--batch-size', default=1, type=int)
    # Dataset parameters
    parser.add_argument('--format', default='fp16', type=str)
    parser.add_argument('--validation', action='store_true', default=False)
    parser.add_argument('--data-path', default='imagenet-div50', type=str, help='dataset path')
    parser.add_argument('--num_workers', default=2, type=int)
    # Benchmark parameters
    parser.add_argument('--only-test', default='', type=str, help='only test a certain model series')
    parser.add_argument('--device', default='CPU', type=str, help='backend device name')
    parser.add_argument('--threads', default=1, type=int, help='number of cpu threads')

    return parser

def benchmark(infer_request, image):
    # warmup
    start = time.perf_counter()
    while time.perf_counter() - start < WARMUP_SEC:
        infer_request.infer(image)

    output = next(iter(infer_request.results.values()))
    val, idx = torch.Tensor(output[0]).topk(3)
    print(list(zip(idx.tolist(), val.tolist())))

    time_list = []
    while sum(time_list) < TEST_SEC:
        start = time.perf_counter()
        infer_request.infer(image)
        time_list.append(time.perf_counter() - start)

    time_max = max(time_list) * 1000
    time_min = min(time_list) * 1000
    time_mean   = np.mean(time_list)   * 1000
    time_median = np.median(time_list) * 1000
    print("min = {:7.2f}ms  max = {:7.2f}ms  mean = {:7.2f}ms, median = {:7.2f}ms".format(time_min, time_max, time_mean, time_median))

def evaluate(data_loader, infer_request, args):
    criterion = torch.nn.CrossEntropyLoss()

    metric_logger = MetricLogger(delimiter="  ")
    header = 'Test:'

    dataset_scale = 50000//args.len_dataset_val
    print(dataset_scale)
    for images, target in metric_logger.log_every(data_loader, 50, header):
        target = target * dataset_scale + (15 if dataset_scale == 50 else 0)

        infer_request.infer(images)
        output = next(iter(infer_request.results.values()))
        output = torch.Tensor(output)

        loss = criterion(output, target)
        acc1, acc5 = accuracy(output, target, topk=(1, 5))
        metric_logger.update(loss=loss.item())

        batch_size = images.shape[0]
        metric_logger.meters['acc1'].update(acc1.item(), n=batch_size)
        metric_logger.meters['acc5'].update(acc5.item(), n=batch_size)

    # gather the stats from all processes
    metric_logger.synchronize_between_processes()
    print('* Acc@1 {top1.global_avg:.3f} Acc@5 {top5.global_avg:.3f} loss {losses.global_avg:.3f}'
          .format(top1=metric_logger.acc1, top5=metric_logger.acc5, losses=metric_logger.loss))
    print(output.mean().item(), output.std().item())

    test_stats = {k: meter.global_avg for k, meter in metric_logger.meters.items()}
    print(f"Accuracy on {args.len_dataset_val} test images: {test_stats['acc1']:.1f}%")

if __name__ == '__main__':
    parser = get_args_parser()
    args = parser.parse_args()

    for name, resolution, usi_eval in [
        ('efficientformerv2_s0', 224, False),
        ('efficientformerv2_s1', 224, False),
        ('efficientformerv2_s2', 224, False),

        ('SwiftFormer_XS', 224, False),
        ('SwiftFormer_S' , 224, False),
        ('SwiftFormer_L1', 224, False),

        ('EMO_1M', 224, False),
        ('EMO_2M', 224, False),
        ('EMO_5M', 224, False),
        ('EMO_6M', 224, False),

        ('edgenext_xx_small', 256, False),
        ('edgenext_x_small' , 256, False),
        ('edgenext_small'   , 256, True),

        ('mobilevitv2_050', 256, False),
        ('mobilevitv2_075', 256, False),
        ('mobilevitv2_100', 256, False),
        # ('mobilevitv2_125', 256, False),
        # ('mobilevitv2_150', 256, False),
        # ('mobilevitv2_175', 256, False),
        # ('mobilevitv2_200', 256, False),

        ('mobilevit_xx_small', 256, False),
        ('mobilevit_x_small' , 256, False),
        ('mobilevit_small'   , 256, False),

        ('LeViT_128S', 224, False),
        ('LeViT_128' , 224, False),
        ('LeViT_192' , 224, False),
        # ('LeViT_256' , 224, False),

        ('resnet50', 224, False),
        ('mobilenetv3_large_100', 224, False),
        ('tf_efficientnetv2_b0' , 224, False),
        ('tf_efficientnetv2_b1' , 240, False),
        ('tf_efficientnetv2_b2' , 260, False),
        # ('tf_efficientnetv2_b3' , 300, False),
    ]:
        if args.only_test and args.only_test not in name and args.only_test != 'ALL' and not args.extern_model:
            continue

        if args.extern_model:
            name = args.extern_model.split(',')[0]
            resolution = int(args.extern_model.split(',')[1])

        args.model = name
        args.input_size = resolution
        args.usi_eval = usi_eval
        print(f"Load openvino model: {name}")
        # Optimize for latency. Most of the devices are configured for latency by default,
        # but there are exceptions like GNA
        config = {}
        config['PERFORMANCE_HINT'] = 'LATENCY'
        if args.format == 'fp32':
            config['INFERENCE_PRECISION_HINT'] = 'f32'
            args.format = 'fp16'
        else:
            config['INFERENCE_PRECISION_HINT'] = 'f16'
        if args.device == 'CPU':
            config['INFERENCE_NUM_THREADS'] = args.threads

        core = ov.Core()
        compiled_model = core.compile_model(f'.xml/{args.format}/{name}.xml', args.device, config)
        ireq = compiled_model.create_infer_request()

        if args.validation:
            dataset_val = build_dataset(args)
            sampler_val = torch.utils.data.SequentialSampler(dataset_val)
            data_loader_val = torch.utils.data.DataLoader(
                dataset_val,
                sampler=sampler_val,
                batch_size=args.batch_size,
                num_workers=args.num_workers,
                drop_last=False
            )
            args.len_dataset_val = len(dataset_val)
            evaluate(data_loader_val, ireq, args)
        else:
            benchmark(ireq, load_image(args))

        if args.extern_model: break

File: src/openvino-perf.cpp
/* Reference code:
   https://github.com/openvinotoolkit/openvino/blob/master/samples/cpp/benchmark/sync_benchmark/main.cpp
   https://github.com/openvinotoolkit/openvino/tree/master/samples/cpp/classification_sample_async
   https://github.com/openvinotoolkit/openvino/blob/master/samples/cpp/hello_query_device/README.md
   https://docs.openvino.ai/2024/openvino-workflow/running-inference/inference-devices-and-modes.html
   https://docs.openvino.ai/2024/openvino-workflow/model-optimization.html
*/

#include <iostream>
#include <numeric>
#include <string>
#include <vector>
#include <iomanip>
#include <filesystem>
#include <getopt.h>
#include <cstring>

#include <openvino/openvino.hpp>
#include <openvino/runtime/intel_npu/properties.hpp>
#include "utils.h"
#include <chrono>
using namespace std::chrono;

struct {
  std::string model;
  bool validation;
  int input_size;
  int batch_size;
  std::string data_path;
} args;

#define USE_OPENVINO
#include "evaluate.tcc"
#include "benchmark.tcc"

int main(int argc, char* argv[])
{
    args.data_path = "imagenet-div50";
    args.validation = false;
    args.batch_size = 1;
    std::string device_name;
    char* arg_long = nullptr;
    char* only_test = nullptr;
    char* extern_model = nullptr;
    int extern_size = 224;
    int num_threads = 1;
    int fpbits = 32;

    static struct option long_options[] =
    {
        {"validation", no_argument, 0, 'v'},
        {"model", required_argument, 0, 'm'},
        {"size", required_argument, 0, 's'},
        {"fp", required_argument, 0, 'f'},
        {"backend",  required_argument, 0, 'u'},
        {"batch-size", required_argument, 0, 'b'},
        {"data-path",  required_argument, 0, 'd'},
        {"only-test",  required_argument, 0, 'o'},
        {"threads",  required_argument, 0, 't'},
        {0, 0, 0, 0}
    };
    int option_index;
    int c;
    while ((c = getopt_long(argc, argv, "vgmsfubot", // TODO
            long_options, &option_index)) != -1)
    {
        switch (c)
        {
            case 0:
            {
                std::cout << "Got long option " << long_options[option_index].name << "." << std::endl;
                arg_long = optarg;
                if (arg_long)
                {
                    std::cout << arg_long << std::endl;
                }
                break;
            }
            case 'm':
                extern_model = optarg;
                break;
            case 's':
                extern_size = atoi(optarg);
                break;
            case 'v':
                args.validation = true;
                break;
            case 'b':
                args.batch_size = atoi(optarg);
                break;
            case 'd':
                args.data_path = optarg;
                break;
            case 'o':
                only_test = optarg;
                break;
            case 'u':
                if (optarg[0] == 'g') {
                    device_name = "GPU";
                    std::cout << "INFO: Using GPU backend" << std::endl;
                }
                else if (optarg[0] == 'n') {
                    device_name = "NPU";
                    std::cout << "INFO: Using NPU backend" << std::endl;
                }
                else {
                    device_name = "CPU";
                    std::cout << "INFO: Using CPU backend" << std::endl;
                }
                break;
            case 't':
                num_threads = atoi(optarg);
                break;
            case 'f':
                fpbits = atoi(optarg);
                break;
            case '?':
                std::cout << "Got unknown option." << std::endl;
                break;
            default:
                std::cout << "Got unknown parse returns: " << c << std::endl;
        }
    }
    std::cout << "INFO: Using num_threads == " << num_threads << std::endl;

    ov::AnyMap device_config = {};
    // Mutable: PERFORMANCE_HINT : LATENCY
    // Mutable: EXECUTION_MODE_HINT : PERFORMANCE
    // Optimize for latency. Most of the devices are configured for latency by default,
    // but there are exceptions like GNA
    device_config.insert(ov::hint::performance_mode(ov::hint::PerformanceMode::LATENCY));
    if (device_name == "CPU") {
        // device_config.insert(ov::hint::enable_hyper_threading(true)); // Damn, worse performace!
        device_config.insert(ov::inference_num_threads(num_threads));
        // device_config.insert(ov::hint::scheduling_core_type(ov::hint::SchedulingCoreType::PCORE_ONLY)); // ECORE_ONLY
    }
    else if (device_name == "NPU") {
        // https://docs.openvino.ai/2024/openvino-workflow/running-inference/inference-devices-and-modes/npu-device.html
        // TODO: has no difference?!
        device_config.insert(ov::intel_npu::compilation_mode_params(ov::Any("optimization-level=2 performance-hint-override=latency")));
        device_config.insert(ov::intel_npu::turbo(true));
    }
    device_config.insert(ov::hint::inference_precision(fpbits == 16? ov::element::f16 : ov::element::f32));

    for (const auto & model: test_models) {
        if (extern_model) {
            args.model = extern_model;
            args.input_size = extern_size;
        }
        else {
            args.model = model.first;
            args.input_size = model.second;
        }
        if (only_test && strcmp(only_test, "ALL") && args.model.find(only_test) == std::string::npos) {
            continue;
        }

        std::cout << "Creating OpenVino infer request: " << args.model << std::endl;
        std::string model_file = ".xml/" + args.model + ".xml";
        if (model_exists(model_file) == 0) {
            std::cerr << args.model << " model doesn't exist!!!" << std::endl;
            continue;
        }
        // Create ov::Core and use it to compile a model.
        // Select the device by providing the name as the second parameter to CLI.
        // Using MULTI device is pointless in sync scenario
        // because only one instance of ov::InferRequest is used
        ov::Core core;
        ov::CompiledModel compiled_model = core.compile_model(model_file, device_name, device_config);
        ov::InferRequest ireq = compiled_model.create_infer_request();
        ov::Tensor input_tensor = ireq.get_input_tensor();

        if (args.validation) {
            evaluate(ireq, input_tensor);
        }
        else {
            benchmark(ireq, input_tensor);
        }

        if (extern_model) {
            break;
        }
    }
}
